apps/language_server/test/experimental/provider/code_action/replace_remote_function_test.exs

```diff
@@ -0,0 +1,322 @@
+    {_, _, action} = code_action("", "/project/file.ex", 0)
+    assert [] = apply(action)
+  end
+
+  test "produces no results if the diagnostic message doesn't fit the format" do
+    assert {_, _, action} =
+             code_action("", "/project/file.ex", 0, diagnostic_message: "This isn't cool")
+
+    assert [] = apply(action)
+  end
+
+  test "produces no results for buggy source code" do
+    {_, _, action} =
+      ~S[
+        1 + 2~/3 ; 4ab(
+        ]
+      |> code_action("/project/file.ex", 0)
+
+    assert [] = apply(action)
+  end
+
+  test "handles nil context" do
+    assert {_, _, action} = code_action("other_var = 6", "/project/file.ex", 0)
+
+    action = put_in(action, [:context], nil)
+
+    assert [] = apply(action)
+  end
+
+  test "handles nil diagnostics" do
+    assert {_, _, action} = code_action("other_var = 6", "/project/file.ex", 0)
+
+    action = put_in(action, [:context, :diagnostics], nil)
+
+    assert [] = apply(action)
+  end
+
+  test "handles empty diagnostics" do
+    assert {_, _, action} = code_action("other_var = 6", "/project/file.ex", 0)
+
+    action = put_in(action, [:context, :diagnostics], [])
+
+    assert [] = apply(action)
+  end
+
+  test "applied to an isolated function" do
+    actual_code = ~S[
+        Enum.counts(a)
+      ]
+
+    expected_doc = ~S[
+        Enum.count(a)
+      ] |> Document.new()
+
+    assert expected_doc ==
+             actual_code
+             |> code_action("/project/file.ex", 1)
+             |> apply_selected_action(0)
+
+    expected_doc = ~S[
+        Enum.concat(a)
+      ] |> Document.new()
+
+    assert expected_doc ==
+             actual_code
+             |> code_action("/project/file.ex", 1)
+             |> apply_selected_action(1)
+  end
+
+  test "works for a function assigned to a variable" do
+    actual_code = ~S[
+      var = &Enum.counts/1
+    ]
+
+    expected_doc = ~S[
+      var = &Enum.count/1
+    ] |> Document.new()
+
+    assert expected_doc ==
+             actual_code
+             |> code_action("/project/file.ex", 1)
+             |> apply_selected_action(0)
+
+    expected_doc = ~S[
+      var = &Enum.concat/1
+    ] |> Document.new()
+
+    assert expected_doc ==
+             actual_code
+             |> code_action("/project/file.ex", 1)
+             |> apply_selected_action(1)
+  end
+
+  test "works with multiple lines" do
+    actual_code = ~S[
+      defmodule MyModule do
+        def my_func(a) do
+          Enum.counts(a)
+        end
+      end
+    ]
+
+    expected_doc = ~S[
+      defmodule MyModule do
+        def my_func(a) do
+          Enum.count(a)
+        end
+      end
+    ] |> Document.new()
+
+    assert expected_doc ==
+             actual_code
+             |> code_action("/project/file.ex", 3)
+             |> apply_selected_action(0)
+
+    expected_doc = ~S[
+      defmodule MyModule do
+        def my_func(a) do
+          Enum.concat(a)
+        end
+      end
+    ] |> Document.new()
```

Comment:
wow, this works very well; I think the structural == will even work with documents. Very nice.